### PR TITLE
aya,aya-obj,integration-test: add better support in `ProgramInfo` & `MapInfo` for old kernels

### PR DESCRIPTION
**Pull request by tyrone-wu**
This adds support for the field availability in `ProgramInfo` and `MapInfo`.

Srry this is long, let me know and I can try to TL;DR it.

For `ProgramInfo`:
- Have `program_type()` to return `bpf_prog_type` instead of the integer representation.
- For fields that can't be `0`, return `Option<NonZero*>` type to indicate `None` as field being unavailable.
- Have `name_as_str()` use the `bpf_name()` feature probe as an additional check for whether field is available. Besides invalid unicode, `None` is returned if probe does not detect feature.
- For `mad_ids()`:
  - Now returns a result of `Option<Vec<NonZeroU32>>`, with `None` indicating field is unavailable.
  - On kernels below v4.15, the init closure that fills in the map info causes an `E2BIG` error since `check_uarg_tail_zero()` expects the entire struct to be zero bytes. I initially worked around this by having `bpf_prog_get_info_by_fd()` do an additional syscall if the first failed with `E2BIG`.
  - Now, `map_ids()` and `bpf_prog_get_info_by_fd()` both uses a new feature probe `prog_info_map_ids()` to detect if field available.
- For `gpl_compatible()`, it now returns `Option<bool>`, with `None` indicating field is unavailable. It uses a new feature probe `prog_info_gpl_compatible()` to detect for field availability.

There was a redundant export of `loaded_programs()` which gave it 2 method of access: `aya::loaded_programs` and `aya::programs::loaded_programs`. To avoid confusion, it should now only be `aya::programs::loaded_programs`.

For `MapInfo`:
- Have `map_type()` to return `bpf_map_type` instead of the integer representation.
- For fields that can't be `0`, return `Option<NonZero*>` type to indicate `None` as field being unavailable.
- For `name_as_str()`, it uses the `bpf_name()` feature probe to detect for field availability. Although the probe performs the check on program name, map name was introduced in the same commit as program name so it should suffice.

Integration tests for `ProgramInfo` and `MapInfo` are updated to use `kernel_assert`/`kernel_assert_eq` macro.
The macro first performs the assertion check. If assertion passes, continue with test. If failed, then check host's kernel version.
If host is below the feature version, then continue with test and log it in stderr (can be observed with `-- --nocapture`). Otherwise, panic 😰.

I verified that this passes on versions (i use ubuntu mainline images for my integration tests):
- 4.13.0
- 4.15.0
- 5.15.0
- 6.1.0

### CODE REVIEW

File: aya-obj/src/links.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            x if x == BPF_LINK_TYPE_NETNS as u32 => BPF_LINK_TYPE_NETNS,
+            x if x == BPF_LINK_TYPE_XDP as u32 => BPF_LINK_TYPE_XDP,
+            x if x == BPF_LINK_TYPE_PERF_EVENT as u32 => BPF_LINK_TYPE_PERF_EVENT,
+            x if x == BPF_LINK_TYPE_KPROBE_MULTI as u32 => BPF_LINK_TYPE_KPROBE_MULTI,
+            x if x == BPF_LINK_TYPE_STRUCT_OPS as u32 => BPF_LINK_TYPE_STRUCT_OPS,
+            x if x == BPF_LINK_TYPE_NETFILTER as u32 => BPF_LINK_TYPE_NETFILTER,
+            x if x == BPF_LINK_TYPE_TCX as u32 => BPF_LINK_TYPE_TCX,
+            x if x == BPF_LINK_TYPE_UPROBE_MULTI as u32 => BPF_LINK_TYPE_UPROBE_MULTI,
+            x if x == BPF_LINK_TYPE_NETKIT as u32 => BPF_LINK_TYPE_NETKIT,
+            _ => __MAX_BPF_LINK_TYPE,
```
Comment (alessandrod) [inline aya-obj/src/links.rs:22]: Technically this is not correct? Looks like this should be try_from instead?

File: aya/src/maps/info.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    pub fn from_id(id: u32) -> Result<Self, MapError> {
+        bpf_map_get_fd_by_id(id)
+            .map_err(MapError::from)
+            .and_then(|fd| Self::new_from_fd(fd.as_fd()))
+    }
+
+    /// The map type as defined by the linux kernel enum [`bpf_map_type`].
+    ///
+    /// Introduced in kernel v4.13.
+    pub fn map_type(&self) -> bpf_map_type {
```
Comment (alessandrod) [inline aya/src/maps/info.rs:44]: we can't re-export generated types as part of the public API, or we can never
guarantee stability. I think we need our own EbpfMapType/EbpfProgramType this
issue has come up before.

File: aya/src/programs/info.rs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+impl ProgramInfo {
+    pub(crate) fn new_from_fd(fd: BorrowedFd<'_>) -> Result<Self, ProgramError> {
+        let info = bpf_prog_get_info_by_fd(fd, &mut [])?;
+        Ok(Self(info))
+    }
+
+    /// The program type as defined by the linux kernel enum [`bpf_prog_type`].
+    ///
+    /// Introduced in kernel v4.13.
+    pub fn program_type(&self) -> bpf_prog_type {
```
Comment (alessandrod) [inline aya/src/programs/info.rs:39]: same here, can't re-export

**Review by alessandrod — COMMENTED** [review]
Thanks for doing this work! Overall looks great, left some comments

File: aya/src/maps/info.rs (hunk 2 above)
Comment (tyrone-wu) [reply aya/src/maps/info.rs:44]: Could I get your thoughts on which would make the most sense for the enum name: 😅
- `EbpfProgramType`
- `EbpfProgram`
- `ProgramType`

I'm not sure if your feedback from https://github.com/aya-rs/aya/pull/959#discussion_r1716884645 applies here, since the `aya::programs::Program` enum is already being used to represent the variants of the actual program implementations.

Personally, I'm kinda leaning more towards `EbpfProgramType` or `ProgramType`, since I feel the `Type` suffix would give better distinction from the `aya::programs::Program` enum.

File: aya-obj/src/links.rs (hunk 1 above)
Comment (tyrone-wu) [reply aya-obj/src/links.rs]: You're right, it should be try_from. 👍

File: aya-obj/src/links.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            x if x == BPF_TCX_EGRESS as u32 => BPF_TCX_EGRESS,
+            x if x == BPF_TRACE_UPROBE_MULTI as u32 => BPF_TRACE_UPROBE_MULTI,
+            x if x == BPF_CGROUP_UNIX_CONNECT as u32 => BPF_CGROUP_UNIX_CONNECT,
+            x if x == BPF_CGROUP_UNIX_SENDMSG as u32 => BPF_CGROUP_UNIX_SENDMSG,
+            x if x == BPF_CGROUP_UNIX_RECVMSG as u32 => BPF_CGROUP_UNIX_RECVMSG,
+            x if x == BPF_CGROUP_UNIX_GETPEERNAME as u32 => BPF_CGROUP_UNIX_GETPEERNAME,
+            x if x == BPF_CGROUP_UNIX_GETSOCKNAME as u32 => BPF_CGROUP_UNIX_GETSOCKNAME,
+            x if x == BPF_NETKIT_PRIMARY as u32 => BPF_NETKIT_PRIMARY,
+            x if x == BPF_NETKIT_PEER as u32 => BPF_NETKIT_PEER,
+            _ => __MAX_BPF_ATTACH_TYPE,
```
[the remark this answers is not in the document]
Comment (tyrone-wu) [reply aya-obj/src/links.rs]: They should be update to `TryFrom` now. 🫡

File: aya/src/maps/info.rs [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+    /// Introduced in kernel v6.9.
+    #[doc(alias = "BPF_MAP_TYPE_ARENA")]
+    Arena = bpf_map_type::BPF_MAP_TYPE_ARENA as isize,
+    /// An invalid map type.
+    #[doc(hidden)]
+    Invalid = bpf_map_type::__MAX_BPF_MAP_TYPE as isize,
+}
+
+impl From<bpf_map_type> for MapType {
+    fn from(map_type: bpf_map_type) -> Self {
```
Comment (alessandrod) [inline aya/src/maps/info.rs:370]: shouldn't this by try_from too?

File: aya/src/maps/info.rs [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+    #[doc(alias = "BPF_MAP_TYPE_CGRP_STORAGE")]
+    CgrpStorage = bpf_map_type::BPF_MAP_TYPE_CGRP_STORAGE as isize,
+    /// An Arena map type.
+    ///
+    /// Introduced in kernel v6.9.
+    #[doc(alias = "BPF_MAP_TYPE_ARENA")]
+    Arena = bpf_map_type::BPF_MAP_TYPE_ARENA as isize,
+    /// An invalid map type.
+    #[doc(hidden)]
+    Invalid = bpf_map_type::__MAX_BPF_MAP_TYPE as isize,
```
Comment (alessandrod) [inline aya/src/maps/info.rs:366]: If we do TryFrom then I don't think we need this?

File: aya/src/programs/info.rs [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+    ///
+    /// Introduced in kernel v6.4.
+    #[doc(alias = "BPF_PROG_TYPE_NETFILTER")]
+    Netfilter = bpf_prog_type::BPF_PROG_TYPE_NETFILTER as isize,
+    /// An invalid program type.
+    #[doc(hidden)]
+    Invalid = bpf_prog_type::__MAX_BPF_PROG_TYPE as isize,
+}
+
+impl From<bpf_prog_type> for ProgramType {
```
Comment (alessandrod) [inline aya/src/programs/info.rs:484]: Same here probably TryFrom?

File: aya/src/programs/info.rs [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+    #[doc(alias = "BPF_PROG_TYPE_SYSCALL")]
+    Syscall = bpf_prog_type::BPF_PROG_TYPE_SYSCALL as isize,
+    /// A Netfilter program type.
+    ///
+    /// Introduced in kernel v6.4.
+    #[doc(alias = "BPF_PROG_TYPE_NETFILTER")]
+    Netfilter = bpf_prog_type::BPF_PROG_TYPE_NETFILTER as isize,
+    /// An invalid program type.
+    #[doc(hidden)]
+    Invalid = bpf_prog_type::__MAX_BPF_PROG_TYPE as isize,
```
Comment (alessandrod) [inline aya/src/programs/info.rs:481]: 🪓

File: aya/src/maps/info.rs (hunk 2 above)
Comment (alessandrod) [reply aya/src/maps/info.rs]: ProgramType and MapType look great :)

**Review by alessandrod — COMMENTED** [review]
Super solid work as always, just a couple of comments then we're ready to go :)

File: aya/src/programs/info.rs (hunk 8 above)
Comment (tyrone-wu) [reply aya/src/programs/info.rs]: 🫡

File: aya/src/maps/info.rs (hunk 6 above)
Comment (tyrone-wu) [reply aya/src/maps/info.rs:366]: 🫡

File: aya/src/programs/info.rs (hunk 7 above)
Comment (tyrone-wu) [reply aya/src/programs/info.rs]: 🫡

File: aya/src/maps/info.rs (hunk 5 above)
Comment (tyrone-wu) [reply aya/src/maps/info.rs]: You're right, `try_from` would probably be more appropriate now that you mentioned `Invalid` won't be needed. 🫡

File: aya/src/maps/info.rs (hunk 6 above)
Comment (alessandrod) [reply aya/src/maps/info.rs:366]: I think you forgot to remove this?

File: aya-obj/src/obj.rs [hunk 9]
```diff
@@ -987,6 +1006,14 @@ pub enum ParseError {
     NoBTF,
 }
 
+/// Invalid bindings to the bpf type from the parsed/received value.
+pub struct InvalidTypeBinding<T> {
+    /// The bpf type to bind to.
+    pub binding_type: String,
```
Comment (alessandrod) [inline aya-obj/src/obj.rs:1012]: This is not being read anywhere I think, so I would remove it? Doesn't matter
much because in practice i don't think it's ever instantiated, but avoid
allocations etc :P

Comment (tyrone-wu) [reply aya-obj/src/obj.rs]: Make sense. 🫡

File: aya/src/maps/info.rs (hunk 6 above)
Comment (tyrone-wu) [reply aya/src/maps/info.rs]: Yep, forgot to 🪓 that. 😓

**Review by alessandrod — APPROVED** [review]

File: aya-log/src/lib.rs [hunk 10]
```diff
@@ -137,19 +136,21 @@ impl EbpfLogger {
...
         let program_info = loaded_programs()
             .filter_map(|info| info.ok())
-            .find(|info| info.id() == program_id)
+            .find(|info| info.id().is_some_and(|id| id.get() == program_id))
             .ok_or(Error::ProgramNotFound)?;
+
         let map = program_info
             .map_ids()
             .map_err(Error::ProgramError)?
+            .expect("`map_ids` field in `bpf_prog_info` not available")
```
Comment (tamird) [inline aya-log/src/lib.rs:145]: why is it ok to panic here?

File: aya-log/src/lib.rs [hunk 11]
```diff
@@ -137,19 +136,21 @@ impl EbpfLogger {
...
             .iter()
-            .filter_map(|id| MapInfo::from_id(*id).ok())
+            .filter_map(|id| MapInfo::from_id(id.get()).ok())
             .find(|map_info| match map_info.name_as_str() {
                 Some(name) => name == MAP_NAME,
                 None => false,
             })
             .ok_or(Error::MapNotFound)?;
-        let map = MapData::from_id(map.id()).map_err(Error::MapError)?;
+        let map = MapData::from_id(map.id().unwrap().get()).map_err(Error::MapError)?;
```
Comment (tamird) [inline aya-log/src/lib.rs:153]: why is it ok to unwrap here?

File: test/integration-test/src/utils.rs [hunk 12]
```diff
@@ -70,3 +70,62 @@ impl Drop for NetNsGuard {
...
+                panic!(
+                    r#"  assertion `left == right` failed: expected host kernel v{current} to have v{feat_version} feature
+    left: {:?}
+   right: {:?}"#,
+                    $left, $right,
+                );
+            } else {
+                // Continue with tests since host is not expected to have feat
+                eprintln!(
+                    r#"ignoring assertion at {}:{}
```
Comment (tamird) [inline test/integration-test/src/utils.rs:120]: we shouldn't ignore the assertion - we should instead assert that the feature is indeed missing.

File: aya/src/programs/info.rs [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+    /// Introduced in kernel v4.15.
+    pub fn map_ids(&self) -> Result<Option<Vec<NonZeroU32>>, ProgramError> {
+        if FEATURES.prog_info_map_ids() {
+            let mut map_ids = vec![0u32; self.0.nr_map_ids as usize];
+            bpf_prog_get_info_by_fd(self.fd()?.as_fd(), &mut map_ids)?;
+
+            Ok(Some(
+                map_ids
+                    .into_iter()
+                    .map(|id| NonZeroU32::new(id).unwrap())
```
Comment (tamird) [inline aya/src/programs/info.rs:133]: this should be .expect with a string that mentions the version detection above.

File: aya-log/src/lib.rs (hunk 10 above)
Comment (tyrone-wu) [reply aya-log/src/lib.rs:145]: Sorry, I definitely shouldn't be panic in the library.

File: aya-log/src/lib.rs (hunk 11 above)
Comment (tyrone-wu) [reply aya-log/src/lib.rs:153]: Sorry, I should have mapped that into an error. I'll have that resolved soon.

File: aya-log/src/lib.rs [hunk 14]
```diff
@@ -137,19 +136,21 @@ impl EbpfLogger {
     ) -> Result<EbpfLogger, Error> {
         let program_info = loaded_programs()
             .filter_map(|info| info.ok())
-            .find(|info| info.id() == program_id)
+            .find(|info| info.id().is_some_and(|id| id.get() == program_id))
```
[the remark this answers is not in the document]
Comment (tyrone-wu) [reply aya-log/src/lib.rs:139]: Ahhhhh you're right. 😓

File: aya/src/programs/info.rs (hunk 13 above)
Comment (tyrone-wu) [reply aya/src/programs/info.rs:133]: Now that I think about it, I think I might of turned the API into an unwrapping nightmare with all the `Option<NonZero*>`. 😵

I'm currently thinking of having fields that were introduced in the same kernel version as the struct not be options (just the direct value). What do you think? 

For `map_ids()`, it looks like it's reasonable to assume the map id would never be zero based on what I traced:
- https://elixir.bootlin.com/linux/v4.15/source/kernel/bpf/syscall.c#L1590: how map IDs are getting filled in `bpf_prog_get_info_by_fd()`
- https://elixir.bootlin.com/linux/v4.15/source/kernel/bpf/verifier.c#L4735: how map info is filled from `bpf_verifier_env` into `bpf_prog_aux`
- https://elixir.bootlin.com/linux/v4.15/source/kernel/bpf/verifier.c#L4249: how map is set in `bpf_verifier_env` before it's verified

Some of the code is shifted in kernel v6.10, but the logic should still be the same. https://elixir.bootlin.com/linux/v6.10/source/kernel/bpf/verifier.c#L18565

cc'ing @alessandrod for thoughts as well

File: test/integration-test/src/utils.rs (hunk 12 above)
Comment (tyrone-wu) [reply test/integration-test/src/utils.rs:120]: So I did some experimenting with `bpf_obj_get_info_by_fd()` after examining the code some more. I tested that I can use `bpf_check_uarg_tail_zero()` to check if a field gives `E2BIG`, however, it would require exposing `bpf_obj_get_info_by_fd()` for the `test` crate to access.

File: aya/src/programs/info.rs (hunk 13 above)
Comment (alessandrod) [reply aya/src/programs/info.rs:133]: > I'm currently thinking of having fields that were introduced in the same kernel version as the struct not be options (just the direct value). What do you think?

Yep this sounds good.

> For map_ids(), it looks like it's reasonable to assume the map id would never be zero based on what I traced:

I don't think that any bpf ids can be 0 regardless of type (map, program, etc)?

In fact I'd probably change all these to return Option<u32>. I think NonZero* is nice to accept as input to enforce safety, but as a return value in these utils seems more annoying than anything else.

Comment (marysaka) [reply aya/src/programs/info.rs:133]: > I don't think that any bpf ids can be 0 regardless of type (map, program, etc)?

It should not be possible yes

> In fact I'd probably change all these to return Option. I think NonZero* is nice to accept as input to enforce safety, but as a return value in these utils seems more annoying than anything else.

Yeah I do agree with you, I would prefer `Option` over `NonZero*` on returns too.

File: aya-obj/src/links.rs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    InvalidTypeBinding,
+};
+
+impl TryFrom<u32> for bpf_link_type {
+    type Error = InvalidTypeBinding<u32>;
+
+    fn try_from(link_type: u32) -> Result<Self, Self::Error> {
+        use bpf_link_type::*;
+        Ok(match link_type {
+            x if x == BPF_LINK_TYPE_UNSPEC as u32 => BPF_LINK_TYPE_UNSPEC,
```
Comment (marysaka) [inline aya-obj/src/links.rs:14]: Kind of a bit nit here but I would instead transform link_type to `bpf_link_type::Type` that would remove all the as u32 boilerplate around :smile:

File: aya-obj/src/links.rs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        })
+    }
+}
+
+impl TryFrom<u32> for bpf_attach_type {
+    type Error = InvalidTypeBinding<u32>;
+
+    fn try_from(attach_type: u32) -> Result<Self, Self::Error> {
+        use bpf_attach_type::*;
+        Ok(match attach_type {
```
Comment (marysaka) [inline aya-obj/src/links.rs:38]: Same comment as earlier but this time with `bpf_attach_type::Type`

File: aya-obj/src/maps.rs [hunk 17]
```diff
@@ -31,7 +25,6 @@ impl TryFrom<u32> for crate::generated::bpf_map_type {
             x if x == BPF_MAP_TYPE_LRU_HASH as u32 => BPF_MAP_TYPE_LRU_HASH,
             x if x == BPF_MAP_TYPE_LRU_PERCPU_HASH as u32 => BPF_MAP_TYPE_LRU_PERCPU_HASH,
             x if x == BPF_MAP_TYPE_LPM_TRIE as u32 => BPF_MAP_TYPE_LPM_TRIE,
-            x if x == BPF_MAP_TYPE_BLOOM_FILTER as u32 => BPF_MAP_TYPE_BLOOM_FILTER,
```
Comment (marysaka) [inline aya-obj/src/maps.rs:34]: Why was that removed?

File: aya-obj/src/maps.rs [hunk 18]
```diff
@@ -42,7 +35,6 @@ impl TryFrom<u32> for crate::generated::bpf_map_type {
             x if x == BPF_MAP_TYPE_CGROUP_STORAGE_DEPRECATED as u32 => {
                 BPF_MAP_TYPE_CGROUP_STORAGE_DEPRECATED
             }
-            x if x == BPF_MAP_TYPE_CGRP_STORAGE as u32 => BPF_MAP_TYPE_CGRP_STORAGE,
```
Comment (marysaka) [inline aya-obj/src/maps.rs:45]: Why was that removed?

File: aya-obj/src/programs/types.rs [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+use crate::{
+    generated::bpf_prog_type::{self, *},
+    InvalidTypeBinding,
+};
+
+impl TryFrom<u32> for bpf_prog_type {
+    type Error = InvalidTypeBinding<u32>;
+
+    fn try_from(prog_type: u32) -> Result<Self, Self::Error> {
+        Ok(match prog_type {
```
Comment (marysaka) [inline aya-obj/src/programs/types.rs:12]: Same comment as earlier but this time with `bpf_prog_type::Type`

File: aya/src/programs/info.rs [hunk 20]
```diff
@@ -0,0 +1,548 @@
...
+    /// Introduced in kernel v4.15.
+    pub fn name_as_str(&self) -> Option<&str> {
+        let name = std::str::from_utf8(self.name()).ok();
+        if let Some(name_str) = name {
+            if FEATURES.bpf_name() || !name_str.is_empty() {
+                return name;
+            }
+        }
+        None
+    }
```
Comment (marysaka) [inline aya/src/programs/info.rs:161]:
```suggestion
    pub fn name_as_str(&self) -> Option<&str> {
        let name = std::str::from_utf8(self.name()).ok()?;

        if FEATURES.bpf_name() || !name.is_empty() {
            return Some(name);
        }

        None
    }
```

File: aya/src/util.rs [hunk 21]
```diff
@@ -177,6 +178,12 @@ impl KernelVersion {
     }
 }
 
+impl Display for KernelVersion {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
+    }
+}
+
```
Comment (marysaka) [inline aya/src/util.rs:186]: Nice thing to see added :+1:

**Review by marysaka — COMMENTED** [review]
Overall, I agree with @alessandrod, we shouldn't use `Option<NonZero*>` as return values but `Option<IntegerType>` instead (0 mapping to None)

Some code pattern could also be simplified I feel here and there but the logic related to those are usually small so that's fine in any cases.

Nice to see all the new comments around kernel version here, I didn't dig too much if the version were appropriate everywhere but if you want I can look more closely at it later :+1:

**Comment by tamird** [comment]
> Overall, I agree with @alessandrod, we shouldn't use `Option<NonZero*>` as return values but `Option<IntegerType>` instead (0 mapping to None)

Why is that better? If we know they can't be zero it's better to communicate that in the type.

**Comment by alessandrod** [comment]
> Why is that better? If we know they can't be zero it's better to communicate that in the type.

Because API ergonomics matter. 

`NonZero<T>` doesn't Deref to T, so I have to call `.get()`.  Having to call get seems worse than having extra information there's no use for, especially since `NonZero<T>::new(v)` return `Option<NonZero<T>>`, so I have to match _and_ get. 

```
// This _does_ provide useful information - I can't call f(0)
fn f(x: NonZeroU32) {
}

// Option tells me whether the map_id is there or not. Then what use is it to me to know that the map id won't be 0? 
// This would be useful if I were to pass the return value to something else, but we have stronger wrappers for things like ProgramIds etc.
fn map_id(&self) -> Option<NonZeroU32> {
}

```

**Comment by tamird** [comment]
> > Why is that better? If we know they can't be zero it's better to communicate that in the type.
> 
> Because API ergonomics matter.
> 
> `NonZero<T>` doesn't Deref to T, so I have to call `.get()`. Having to call get seems worse than having extra information there's no use for, especially since `NonZero<T>::new(v)` return `Option<NonZero<T>>`, so I have to match _and_ get.
> 
> ```
> // This _does_ provide useful information - I can't call f(0)
> fn f(x: NonZeroU32) {
> }
> 
> // Option tells me whether the map_id is there or not. Then what use is it to me to know that the map id won't be 0? 
> // This would be useful if I were to pass the return value to something else, but we have stronger wrappers for things like ProgramIds etc.
> fn map_id(&self) -> Option<NonZeroU32> {
> }
> ```

Then why expose the scalar at all? If the value doesn't matter, make it opaque.

**Comment by alessandrod** [comment]
> Then why expose the scalar at all? If the value doesn't matter, make it opaque.

We should definitely have types for all these IDs, but babysteps. Even without stronger ids this PR was an improvement over what was there before, and removing NonZeroU32 is an improvement over what's there, imo.

File: aya-obj/src/maps.rs (hunk 17 above)
Comment (tyrone-wu) [reply aya-obj/src/maps.rs:34]: There was a duplicate match for `bloom_filter` line 58.

File: aya-obj/src/maps.rs (hunk 18 above)
Comment (tyrone-wu) [reply aya-obj/src/maps.rs:45]: Also duplicate `cgro_storage` match on line 60.

File: aya-obj/src/links.rs (hunk 15 above)
Comment (tyrone-wu) [reply aya-obj/src/links.rs:14]: Could you elaborate on `::Type`? I'm not familiar with the usage of that syntax. 🤔

Comment (marysaka) [reply aya-obj/src/links.rs:14]: Those `BPF_LINK_TYPE_` values have a type of `bpf_link_type::Type` (that you can find defined [here for x86_64](https://github.com/aya-rs/aya/blob/635ed3baed5442c1364a360d7234b72c4ffe3fd8/ebpf/aya-ebpf-bindings/src/x86_64/bindings.rs#L698))

So the idea would be to cast link_type to `bpf_link_type::Type` instead of casting each  `BPF_LINK_TYPE_` values to u32.

Comment (tamird) [reply aya-obj/src/links.rs:14]: @tyrone-wu did you want to address this one?

File: test/integration-test/src/utils.rs (hunk 12 above)
Comment (tamird) [reply test/integration-test/src/utils.rs:120]: That seems OK to me. Can you revisit or file an issue?

File: aya/src/sys/bpf.rs [hunk 22]
```diff
@@ -694,6 +703,62 @@ pub(crate) fn is_prog_name_supported() -> bool {
     bpf_prog_load(&mut attr).is_ok()
 }
 
+/// Tests whether `nr_map_ids` & `map_ids` fields in `bpf_prog_info` is available.
+pub(crate) fn is_info_map_ids_supported() -> bool {
```
[the remark this answers is not in the document]
Comment (tamird) [reply aya/src/sys/bpf.rs:707]: reminder that we might want to deduplicate here.

**Review by tamird — COMMENTED** [review]
Just pinging some leftovers.

File: aya/src/sys/bpf.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (tamird) [reply aya/src/sys/bpf.rs:707]: https://github.com/aya-rs/aya/pull/1165

File: aya-obj/src/links.rs (hunk 15 above)
Comment (tamird) [reply aya-obj/src/links.rs:14]: I think this was done.

File: test/integration-test/src/utils.rs (hunk 12 above)
Comment (tyrone-wu) [reply test/integration-test/src/utils.rs:120]: I can revisit/ 👍

File: aya-log/src/lib.rs (hunk 10 above)
Comment (tamird) [reply aya-log/src/lib.rs:145]: https://github.com/aya-rs/aya/commit/02d1db5fc043fb7af90c14d13de6419ec5b9bcb5

File: aya-log/src/lib.rs (hunk 11 above)
Comment (tamird) [reply aya-log/src/lib.rs:153]: https://github.com/aya-rs/aya/commit/02d1db5fc043fb7af90c14d13de6419ec5b9bcb5

File: test/integration-test/src/utils.rs (hunk 12 above)
Comment (tamird) [reply test/integration-test/src/utils.rs:120]: https://github.com/aya-rs/aya/pull/1219

File: aya/src/programs/info.rs (hunk 13 above)
Comment (tamird) [reply aya/src/programs/info.rs:133]: https://github.com/aya-rs/aya/commit/02d1db5fc043fb7af90c14d13de6419ec5b9bcb5

File: aya-obj/src/links.rs (hunk 15 above)
Comment (tamird) [reply aya-obj/src/links.rs:14]: This wasn't done, but the suggestion is UB. Casting an out-of-range integer to a Rust enum is UB.

File: aya-obj/src/links.rs (hunk 16 above)
Comment (tamird) [reply aya-obj/src/links.rs:38]: Same as above, doing this is UB.
